File: python/mpl_sdk/session.py
"""
MPL Session Management

Provides session handling for MCP/A2A server connections with MPL overlay.
"""

import json
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional

import aiohttp
import websockets

from mpl_sdk._mpl_core import (
    MplEnvelope,
    QomProfile,
    SchemaValidator,
    SType,
    semantic_hash,
)
from mpl_sdk.errors import (
    ConnectionError,
    NegotiationError,
    SchemaFidelityError,
    UnknownStypeError,
)


@dataclass
class SessionConfig:
    """Configuration for an MPL session."""

    endpoint: str
    """MCP/A2A server endpoint URL."""

    stypes: List[str] = field(default_factory=list)
    """List of SType identifiers this session supports."""

    qom_profile: Optional[str] = None
    """QoM profile name to enforce (e.g., 'qom-strict-argcheck')."""

    registry_path: Optional[str] = None
    """Path to local SType registry. Defaults to ./registry."""

    timeout_ms: int = 30000
    """Request timeout in milliseconds."""

    auto_validate: bool = True
    """Automatically validate payloads against schemas."""

    auto_hash: bool = True
    """Automatically compute semantic hashes."""


@dataclass
class NegotiatedCapabilities:
    """Result of AI-ALPN handshake negotiation."""

    common_stypes: List[str]
    """STypes supported by both client and server."""

    selected_profile: Optional[str]
    """QoM profile agreed upon."""

    server_extensions: Dict[str, Any] = field(default_factory=dict)
    """Additional capabilities advertised by server."""


class Session:
    """
    MPL Session for typed communication with MCP/A2A servers.

    Example:
        async with Session(SessionConfig(
            endpoint="ws://localhost:8080/mcp",
            stypes=["org.calendar.Event.v1", "org.agent.TaskPlan.v1"],
            qom_profile="qom-basic",
        )) as session:
            # Send typed request
            response = await session.send(
                stype="org.calendar.Event.v1",
                payload={"title": "Meeting", "start": "2024-01-15T10:00:00Z", ...}
            )
    """

    def __init__(self, config: SessionConfig):
        self.config = config
        self._ws: Optional[websockets.WebSocketClientProtocol] = None
        self._http: Optional[aiohttp.ClientSession] = None
        self._validators: Dict[str, SchemaValidator] = {}
        self._negotiated: Optional[NegotiatedCapabilities] = None
        self._qom_profile: Optional[QomProfile] = None
        self._message_handlers: Dict[str, Callable] = {}
        self._connected = False

    async def __aenter__(self) -> "Session":
        await self.connect()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        await self.close()

    async def connect(self) -> NegotiatedCapabilities:
        """
        Establish connection and perform AI-ALPN handshake.

        Returns:
            NegotiatedCapabilities with common stypes and selected profile.

        Raises:
            ConnectionError: If connection fails.
            NegotiationError: If handshake fails.
        """
        endpoint = self.config.endpoint

        try:
            if endpoint.startswith("ws://") or endpoint.startswith("wss://"):
                await self._connect_websocket()
            else:
                await self._connect_http()

            # Perform AI-ALPN handshake
            self._negotiated = await self._handshake()
            self._connected = True

            # Load QoM profile if specified
            if self.config.qom_profile:
                self._qom_profile = await self._load_qom_profile(self.config.qom_profile)

            return self._negotiated

        except Exception as e:
            raise ConnectionError(
                message=f"Failed to connect to {endpoint}: {e}",
                endpoint=endpoint,
                cause=str(e),
            )

    async def _connect_websocket(self) -> None:
        """Establish WebSocket connection."""
        self._ws = await websockets.connect(
            self.config.endpoint,
            close_timeout=self.config.timeout_ms / 1000,
        )

    async def _connect_http(self) -> None:
        """Establish HTTP session."""
        timeout = aiohttp.ClientTimeout(total=self.config.timeout_ms / 1000)
        self._http = aiohttp.ClientSession(timeout=timeout)

    async def _handshake(self) -> NegotiatedCapabilities:
        """Perform AI-ALPN capability negotiation."""
        client_hello = {
            "type": "ai-alpn-hello",
            "version": "1.0",
            "stypes": self.config.stypes,
            "qom_profiles": [self.config.qom_profile] if self.config.qom_profile else [],
        }

        if self._ws:
            await self._ws.send(json.dumps(client_hello))
            response_text = await self._ws.recv()
            response = json.loads(response_text)
        elif self._http:
            async with self._http.post(
                f"{self.config.endpoint}/.well-known/ai-alpn",
                json=client_hello,
            ) as resp:
                response = await resp.json()
        else:
            raise ConnectionError(
                message="No connection established",
                endpoint=self.config.endpoint,
            )

        if response.get("type") == "ai-alpn-error":
            raise NegotiationError(
                message=response.get("message", "Handshake failed"),
                client_stypes=self.config.stypes,
                server_stypes=response.get("server_stypes", []),
                reason=response.get("reason"),
            )

        return NegotiatedCapabilities(
            common_stypes=response.get("common_stypes", []),
            selected_profile=response.get("selected_profile"),
            server_extensions=response.get("extensions", {}),
        )

    async def _load_qom_profile(self, profile_name: str) -> QomProfile:
        """Load QoM profile by name.

        Currently supports built-in profiles:
        - qom-basic: Schema Fidelity only
        - qom-strict-argcheck: SF + Instruction Compliance
        """
        # Match known profiles
        if profile_name == "qom-strict-argcheck":
            return QomProfile.strict_argcheck()
        else:
            # Default to basic profile for unknown names
            return QomProfile.basic()

    async def send(
        self,
        stype: str,
        payload: Dict[str, Any],
        validate: Optional[bool] = None,
        compute_hash: Optional[bool] = None,
    ) -> MplEnvelope:
        """
        Send a typed payload to the server.

        Args:
            stype: SType identifier (e.g., "org.calendar.Event.v1")
            payload: The payload data matching the SType schema
            validate: Override auto_validate setting
            compute_hash: Override auto_hash setting

        Returns:
            MplEnvelope containing the response

        Raises:
            SchemaFidelityError: If validation fails
            ConnectionError: If not connected
        """
        if not self._connected:
            raise ConnectionError(
                message="Session not connected",
                endpoint=self.config.endpoint,
            )

        # Convert payload to JSON string for processing
        payload_json = json.dumps(payload)

        # Validate against schema
        should_validate = validate if validate is not None else self.config.auto_validate
        if should_validate:
            await self._validate_payload(stype, payload_json)

        # Compute semantic hash
        should_hash = compute_hash if compute_hash is not None else self.config.auto_hash
        sem_hash = None
        if should_hash:
            sem_hash = semantic_hash(payload_json)

        # Create envelope
        envelope = MplEnvelope(
            stype=stype,
            payload=payload_json,
            profile=self.config.qom_profile,
        )
        if sem_hash:
            envelope.sem_hash = sem_hash

        # Send via appropriate transport
        if self._ws:
            await self._ws.send(envelope.to_json())
            response_text = await self._ws.recv()
            response = self._parse_envelope(response_text)
        elif self._http:
            async with self._http.post(
                self.config.endpoint,
                json=json.loads(envelope.to_json()),
            ) as resp:
                response_data = await resp.json()
                response = self._parse_envelope(json.dumps(response_data))
        else:
            raise ConnectionError(
                message="No connection established",
                endpoint=self.config.endpoint,
            )

        # Validate response if auto_validate is enabled
        if should_validate and response.stype:
            await self._validate_payload(response.stype, response.payload)

        return response

    def _parse_envelope(self, json_str: str) -> MplEnvelope:
        """Parse JSON string into MplEnvelope."""
        data = json.loads(json_str)
        envelope = MplEnvelope(
            stype=data.get("stype", ""),
            payload=json.dumps(data.get("payload", {})),
            args_stype=data.get("args_stype"),
            profile=data.get("profile"),
        )
        envelope.sem_hash = data.get("sem_hash")
        return envelope

    async def _validate_payload(self, stype: str, payload_json: str) -> None:
        """Validate payload against SType schema."""
        validator = await self._get_validator(stype)
        result = validator.validate(stype, payload_json)

        if not result.valid:
            raise SchemaFidelityError(
                message=f"Payload does not match schema for {stype}",
                stype=stype,
                validation_errors=[{"path": e.path, "message": e.message} for e in result.errors],
            )

    async def _get_validator(self, stype: str) -> SchemaValidator:
        """Get or create validator for SType."""
        if stype in self._validators:
            return self._validators[stype]

        # Parse SType to find schema path
        parsed = SType(stype)
        registry_path = self.config.registry_path or "./registry"
        schema_path = (
            f"{registry_path}/stypes/{parsed.namespace}/{parsed.domain}/"
            f"{parsed.name}/v{parsed.major_version}/schema.json"
        )

        try:
            with open(schema_path) as f:
                schema_json = f.read()
            validator = SchemaValidator()
            validator.register(stype, schema_json)
            self._validators[stype] = validator
            return validator
        except FileNotFoundError:
            raise UnknownStypeError(stype=stype, registry_path=registry_path)

    def on_message(self, stype: str) -> Callable:
        """
        Decorator to register a handler for incoming messages of a specific SType.

        Example:
            @session.on_message("org.agent.TaskPlan.v1")
            async def handle_task_plan(envelope: MplEnvelope):
                print(f"Received task plan: {envelope.payload}")
        """

        def decorator(func: Callable) -> Callable:
            self._message_handlers[stype] = func
            return func

        return decorator

    async def listen(self) -> None:
        """
        Start listening for incoming messages (WebSocket only).

        Dispatches messages to registered handlers based on SType.
        """
        if not self._ws:
            raise ConnectionError(
                message="listen() requires WebSocket connection",
                endpoint=self.config.endpoint,
            )

        async for message in self._ws:
            try:
                envelope = self._parse_envelope(message)
                handler = self._message_handlers.get(envelope.stype)
                if handler:
                    await handler(envelope)
            except (json.JSONDecodeError, ValueError):
                continue  # Skip malformed messages

    async def close(self) -> None:
        """Close the session and cleanup resources."""
        if self._ws:
            await self._ws.close()
            self._ws = None

        if self._http:
            await self._http.close()
            self._http = None

        self._connected = False
        self._negotiated = None

    @property
    def is_connected(self) -> bool:
        """Check if session is connected."""
        return self._connected

    @property
    def capabilities(self) -> Optional[NegotiatedCapabilities]:
        """Get negotiated capabilities, or None if not connected."""
        return self._negotiated

File: typescript/src/session/session.ts
/**
 * MPL Session - manages connection and typed communication
 */

import WebSocket from 'ws';
import { MplEnvelope, MplEnvelopeOptions } from '../types/envelope';
import { QomProfile, QomMetrics } from '../types/qom';
import { SchemaValidator } from '../validation/schema-validator';
import { semanticHash } from '../validation/hash';
import {
  ConnectionError,
  NegotiationError,
  SchemaFidelityError,
} from '../errors';

export interface SessionConfig {
  /** Server endpoint URL (ws:// or http://) */
  endpoint: string;
  /** List of STypes this session supports */
  stypes?: string[];
  /** QoM profile to enforce */
  qomProfile?: string;
  /** Path to local registry */
  registryPath?: string;
  /** Request timeout in ms */
  timeoutMs?: number;
  /** Auto-validate payloads */
  autoValidate?: boolean;
  /** Auto-compute semantic hashes */
  autoHash?: boolean;
}

export interface NegotiatedCapabilities {
  commonStypes: string[];
  selectedProfile?: string;
  serverExtensions: Record<string, unknown>;
}

export interface SendOptions {
  validate?: boolean;
  computeHash?: boolean;
}

type MessageHandler = (envelope: MplEnvelope) => void | Promise<void>;

export class Session {
  private config: SessionConfig;
  private ws?: WebSocket;
  private validator: SchemaValidator;
  private qomProfile?: QomProfile;
  private negotiated?: NegotiatedCapabilities;
  private connected: boolean = false;
  private messageHandlers: Map<string, MessageHandler> = new Map();
  private pendingRequests: Map<string, {
    resolve: (value: MplEnvelope) => void;
    reject: (error: Error) => void;
  }> = new Map();

  constructor(config: SessionConfig) {
    this.config = {
      timeoutMs: 30000,
      autoValidate: true,
      autoHash: true,
      stypes: [],
      ...config,
    };
    this.validator = new SchemaValidator();
  }

  /**
   * Connect to the server and perform AI-ALPN handshake
   */
  async connect(): Promise<NegotiatedCapabilities> {
    const { endpoint } = this.config;

    try {
      if (endpoint.startsWith('ws://') || endpoint.startsWith('wss://')) {
        await this.connectWebSocket();
      } else {
        throw new ConnectionError(endpoint, 'Only WebSocket connections supported');
      }

      // Perform AI-ALPN handshake
      this.negotiated = await this.handshake();
      this.connected = true;

      // Load QoM profile
      if (this.config.qomProfile) {
        this.qomProfile = this.loadQomProfile(this.config.qomProfile);
      }

      return this.negotiated;
    } catch (error) {
      throw new ConnectionError(
        endpoint,
        error instanceof Error ? error.message : String(error)
      );
    }
  }

  /**
   * Establish WebSocket connection
   */
  private connectWebSocket(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.ws = new WebSocket(this.config.endpoint);

      this.ws.on('open', () => resolve());
      this.ws.on('error', (err) => reject(err));
      this.ws.on('message', (data) => this.handleMessage(data.toString()));
      this.ws.on('close', () => {
        this.connected = false;
      });
    });
  }

  /**
   * Perform AI-ALPN handshake
   */
  private async handshake(): Promise<NegotiatedCapabilities> {
    const clientHello = {
      type: 'ai-alpn-hello',
      version: '1.0',
      stypes: this.config.stypes,
      qom_profiles: this.config.qomProfile ? [this.config.qomProfile] : [],
    };

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('Handshake timeout'));
      }, this.config.timeoutMs);

      const handler = (data: string) => {
        clearTimeout(timeout);
        const response = JSON.parse(data);

        if (response.type === 'ai-alpn-error') {
          reject(
            new NegotiationError(
              this.config.stypes ?? [],
              response.server_stypes ?? [],
              response.message
            )
          );
          return;
        }

        resolve({
          commonStypes: response.common_stypes ?? [],
          selectedProfile: response.selected_profile,
          serverExtensions: response.extensions ?? {},
        });
      };

      // Temporarily override message handler for handshake
      if (this.ws) {
        this.ws.once('message', (data) => handler(data.toString()));
        this.ws.send(JSON.stringify(clientHello));
      }
    });
  }

  /**
   * Load a QoM profile
   */
  private loadQomProfile(profileName: string): QomProfile {
    switch (profileName) {
      case 'qom-strict-argcheck':
        return QomProfile.strictArgcheck();
      case 'qom-outcome':
        return QomProfile.outcome();
      default:
        return QomProfile.basic();
    }
  }

  /**
   * Handle incoming WebSocket message
   */
  private handleMessage(data: string): void {
    try {
      const envelope = MplEnvelope.fromJSON(data);
      const handler = this.messageHandlers.get(envelope.stype);

      if (handler) {
        handler(envelope);
      }

      // Check for pending request response
      const pending = this.pendingRequests.get(envelope.id);
      if (pending) {
        this.pendingRequests.delete(envelope.id);
        pending.resolve(envelope);
      }
    } catch (error) {
      console.error('Failed to handle message:', error);
    }
  }

  /**
   * Register a message handler for an SType
   */
  onMessage(stype: string, handler: MessageHandler): void {
    this.messageHandlers.set(stype, handler);
  }

  /**
   * Send a typed payload
   */
  async send(
    stype: string,
    payload: Record<string, unknown>,
    options: SendOptions = {}
  ): Promise<MplEnvelope> {
    if (!this.connected) {
      throw new ConnectionError(this.config.endpoint, 'Not connected');
    }

    const shouldValidate = options.validate ?? this.config.autoValidate;
    const shouldHash = options.computeHash ?? this.config.autoHash;

    // Validate payload
    if (shouldValidate && this.validator.hasSchema(stype)) {
      const result = this.validator.validate(stype, payload);
      if (!result.valid) {
        throw new SchemaFidelityError(stype, result.errors);
      }
    }

    // Compute semantic hash
    const semHash = shouldHash ? semanticHash(payload) : undefined;

    // Create envelope
    const envelope = new MplEnvelope({
      stype,
      payload,
      profile: this.config.qomProfile,
      semHash,
    });

    // Send and wait for response
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pendingRequests.delete(envelope.id);
        reject(new Error('Request timeout'));
      }, this.config.timeoutMs);

      this.pendingRequests.set(envelope.id, {
        resolve: (response) => {
          clearTimeout(timeout);
          resolve(response);
        },
        reject: (error) => {
          clearTimeout(timeout);
          reject(error);
        },
      });

      this.ws?.send(envelope.toJSON());
    });
  }

  /**
   * Register a schema for validation
   */
  registerSchema(stype: string, schema: object | string): void {
    this.validator.register(stype, schema);
  }

  /**
   * Close the session
   */
  async close(): Promise<void> {
    if (this.ws) {
      this.ws.close();
      this.ws = undefined;
    }
    this.connected = false;
    this.negotiated = undefined;
  }

  /**
   * Check if connected
   */
  get isConnected(): boolean {
    return this.connected;
  }

  /**
   * Get negotiated capabilities
   */
  get capabilities(): NegotiatedCapabilities | undefined {
    return this.negotiated;
  }
}
